src/app/loadout-drawer/loadout-drawer-reducer.ts

```diff
@@ -1,26 +1,59 @@
 import { LoadoutParameters } from '@destinyitemmanager/dim-api-types';
+import { D1Categories } from 'app/destiny1/d1-bucket-categories';
 import { D1ManifestDefinitions } from 'app/destiny1/d1-definitions';
+import { D2Categories } from 'app/destiny2/d2-bucket-categories';
 import { D2ManifestDefinitions } from 'app/destiny2/d2-definitions';
 import { t } from 'app/i18next-t';
 import { DimItem } from 'app/inventory/item-types';
+import { DimStore } from 'app/inventory/store-types';
 import { SocketOverrides } from 'app/inventory/store/override-sockets';
 import { showNotification } from 'app/notifications/notifications';
 import { itemCanBeInLoadout } from 'app/utils/item-utils';
 import { getSocketsByCategoryHash } from 'app/utils/socket-utils';
 import { DestinyClass, TierType } from 'bungie-api-ts/destiny2';
-import { SocketCategoryHashes } from 'data/d2/generated-enums';
+import { BucketHashes, SocketCategoryHashes } from 'data/d2/generated-enums';
 import produce from 'immer';
 import _ from 'lodash';
 import { Loadout, LoadoutItem, ResolvedLoadoutItem } from './loadout-types';
-import { singularBucketHashes } from './loadout-utils';
+import {
+  createSocketOverridesFromEquipped,
+  extractArmorModHashes,
+  fromEquippedTypes,
+  getUnequippedItemsForLoadout,
+  singularBucketHashes,
+} from './loadout-utils';
+
+// TODO: should this really be a reducer, or just a series of functions that produce a new loadout coupled w/ a setLoadout function? General loadout manipulation library would be useful...
+// Each function should return a (loadout: Loadout) => Loadout
```

Comment:
Yeah I think this would be really nice. They could be reused in a bunch of places I would say